[te] Don't fuse integer fmod or remainder

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48700 [te] Don't fuse integer fmod or remainder**

fmod and remainder on int tensors will raise ZeroDivisionError if their divisors are 0.  I don't think we should try to generate code that raises exceptions.  If at some point we really wanted to fuse these, I might lean towards calling a C++ helper function from the generated code.

Differential Revision: [D25265792](https://our.internmc.facebook.com/intern/diff/D25265792/)

bertmaher committed:
[te] Don't fuse integer fmod or remainder

fmod and remainder on int tensors will raise ZeroDivisionError if their divisors are 0.  I don't think we should try to generate code that raises exceptions.  If at some point we really wanted to fuse these, I might lean towards calling a C++ helper function from the generated code.

Differential Revision: [D25265792](https://our.internmc.facebook.com/intern/diff/D25265792/)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -191,6 +199,20 @@ bool isSupported(Node* node) {
+          if (!st || !isFloatingType(*st)) {
+            return false;
+          }
+        } else if (!v->type()->cast<FloatType>()) {
```
This doesn't include Half and Double, does it?

Krovatkin replied:
It might. We don't seem to differentiate statically halfs, doubles and single-points, @eellison ?

eellison replied:
this check is on a non-Tensor type (Scalar value), so Half/Double doesn't apply here

ZolotukhinM commented on test/test_jit_fuser_te.py:
```diff
@@ -1288,7 +1288,6 @@ def apply(fn):
             return lambda x, y: fn(x, y)
 
         dtypes = [
-            # FIXME: Fails in IR Eval: torch.int8 and_ cpu
```
This FIXME is for a different issue.

bertmaher replied:
What's the issue?  It looked like a vestigial FIXME.

ZolotukhinM replied:
The issue is that the combination op=and, dtype=int8, device=cpu fails on IR eval :) I added it in one of the recent commits

bertmaher replied:
Ohhh, so what threw me off was this being in the block of `dtypes`.  I read it as saying the combination of int8 and cpu fails :-p.  Anyways I'll put it back

ZolotukhinM reviewed (commented):
Overall looks good, but I wonder if division by 0 was the only issue. There is another test called `test_binary_div_ops` that doesn't try to divide by 0 but still fails with int8. Only fusing floats would solve that too, but I'm curious what was the problem there - we might have some other bug.

bertmaher commented:
> Overall looks good, but I wonder if division by 0 was the only issue. There is another test called `test_binary_div_ops` that doesn't try to divide by 0 but still fails with int8. Only fusing floats would solve that too, but I'm curious what was the problem there - we might have some other bug.

Actually there are at least two other issues which end up being side-stepped by this PR.  I should probably fix them too, really, although they become more annoying to test if ints are banned in fusion (which I still think is the right call for the ZeroDivisionError reason).
1. We translate fmod(int, float) to (int)fmod(float, float), but eager mode translates it to (int)(int % int).  Banning ints at the fuser level masks this bad translation.
2. Even if we somehow did create TE for fmod(int, int) we throw unsupported_dtype from llvm codegen.  Which isn't *entirely* wrong, since we arguably should translate it as normal mod(int, int).

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -191,6 +199,20 @@ bool isSupported(Node* node) {
+          if (!st || !isFloatingType(*st)) {
+            return false;
+          }
+        } else if (!v->type()->cast<FloatType>()) {
```
You could also allow constant non-zero ints:

`(v->type()->cast<IntType>() && constant_as<int>(v).value_or(0) != 0)`

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -191,6 +199,20 @@ bool isSupported(Node* node) {
       }
     }
 
+    // Value is only supported if operands are floats.
```
Do we need to exclude non-floating point  tensors if they are the first input ?

bertmaher replied:
That's a good point, this could be relaxed.

eellison reviewed (approved):
LGTM

bertmaher committed:
Update on "[te] Don't fuse integer fmod or remainder"


fmod and remainder on int tensors will raise ZeroDivisionError if their divisors are 0.  I don't think we should try to generate code that raises exceptions.  If at some point we really wanted to fuse these, I might lean towards calling a C++ helper function from the generated code.

Differential Revision: [D25265792](https://our.internmc.facebook.com/intern/diff/D25265792/)

[ghstack-poisoned]

bertmaher committed:
Update on "[te] Don't fuse integer fmod or remainder"


fmod and remainder on int tensors will raise ZeroDivisionError if their divisors are 0.  I don't think we should try to generate code that raises exceptions.  If at some point we really wanted to fuse these, I might lean towards calling a C++ helper function from the generated code.

Differential Revision: [D25265792](https://our.internmc.facebook.com/intern/diff/D25265792/)

[ghstack-poisoned]

bertmaher committed:
Update on "[te] Don't fuse integer fmod or remainder"


fmod and remainder on int tensors will raise ZeroDivisionError if their divisors are 0.  I don't think we should try to generate code that raises exceptions.  If at some point we really wanted to fuse these, I might lean towards calling a C++ helper function from the generated code.

Differential Revision: [D25265792](https://our.internmc.facebook.com/intern/diff/D25265792/)

[ghstack-poisoned]

bertmaher committed:
remove now-irrelevant test on "[te] Don't fuse integer fmod or remainder"


fmod and remainder on int tensors will raise ZeroDivisionError if their divisors are 0.  I don't think we should try to generate code that raises exceptions.  If at some point we really wanted to fuse these, I might lean towards calling a C++ helper function from the generated code.

Differential Revision: [D25265792](https://our.internmc.facebook.com/intern/diff/D25265792/)

[ghstack-poisoned]

bertmaher committed:
flake8 on "[te] Don't fuse integer fmod or remainder"


fmod and remainder on int tensors will raise ZeroDivisionError if their divisors are 0.  I don't think we should try to generate code that raises exceptions.  If at some point we really wanted to fuse these, I might lean towards calling a C++ helper function from the generated code.

Differential Revision: [D25265792](https://our.internmc.facebook.com/intern/diff/D25265792/)

[ghstack-poisoned]

bertmaher committed:
Move input typecheck out of isSupported on "[te] Don't fuse integer fmod or remainder"


fmod and remainder on int tensors will raise ZeroDivisionError if their divisors are 0.  I don't think we should try to generate code that raises exceptions.  If at some point we really wanted to fuse these, I might lean towards calling a C++ helper function from the generated code.

Differential Revision: [D25265792](https://our.internmc.facebook.com/intern/diff/D25265792/)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -727,6 +729,34 @@ class TensorExprFuser {
     return true;
   }
 
+  bool typesAreSupported(const Node* node) {
```
While you’re at it, could you move the cuda_only fuse set ?
